### PR TITLE
MODELIX-724: stabilize plugin GUI

### PR DESCRIPTION
**Pull request by benedekh**
Fixes [MODELIX-724](https://issues.modelix.org/issue/MODELIX-724)

### CODE REVIEW

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt [hunk 1]
```diff
@@ -271,64 +274,82 @@ class ModelSyncGuiFactory : ToolWindowFactory, Disposable {
...
-                        moduleModel.removeAllElements()
-                        val children = branch.getRootNode().allChildren.map {
-                            val name = it.getPropertyValue(BuiltinLanguages.jetbrains_mps_lang_core.INamedConcept.name)
-                                ?: it.toString()
-                            INodeWithName(it, name)
-                        }
-                        moduleModel.addAll(children.toList())
+                val client = existingConnectionsModel.selectedItem as ModelClientV2
+                val branchReference = branchModel.selectedItem as BranchReference
+                val branch = client.getReplicatedModel(branchReference).start()
```
Comment (odzhychko) [inline mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt:319]: The `ReplicatedModel` here is not closed (and not used further).
Also fetching all data does not make sense here.

I would suggest to use a ModelQL query here.

Comment (slisson) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt:319]: These replicated models are probably used at other places as well. Their life-cycle should be managed somewhere globally and then being reused.

Comment (odzhychko) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt:319]: Definitely.  If I understood it correctly, the new plugin will only have one connection and one branch checked out per project. So there should be only one ReplicatedModel per project an its lifecycle should be bound to the project.

(I think I would not try to optimize reusing Clients/ReplicatedModels between opened projects to make it simple (and seperate failures).)

---

In this specific case, the user is still just exploring available branches and available modules in the UI.
For that, we definitely should not load the whole model and, for example, use ModelQL instead.

Comment (benedekh) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt]: fixed in [d4bb450](https://github.com/modelix/modelix.mps-plugins/pull/73/commits/d4bb45091b260acb9154445de5ca274ff4cb4c3e)

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt [hunk 2]
```diff
@@ -136,10 +136,13 @@ class SNodeFactory(
             val sourceNodeId = iNode.nodeIdAsLong()
             val source = nodeMap.getNode(sourceNodeId)!!
 
-            val sReferenceLink = (it.first as MPSReferenceLink).link
-            val reference = SReferenceLinkAdapter2(sReferenceLink.id, sReferenceLink.name)
-            val targetNodeId = it.second.nodeIdAsLong()
+            val reference = when (it.first) {
+                is MPSReferenceLink -> (it.first as MPSReferenceLink).link
+                is ReferenceLinkFromName -> source.concept.referenceLinks.first { refLink -> refLink.name == it.first.getSimpleName() }
```
Comment (odzhychko) [inline mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt:141]: A comment on why `ReferenceLinkFromName` can happen.
Or even better a test...

Comment (odzhychko) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt:141]: In general, I'd like to have tests for all new features except Swing-UI code (that should be kept minimal)... Not sure how realistic it is.

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/transformation/modelixToMps/initial/ITreeToSTreeTransformer.kt [hunk 3]
```diff
@@ -18,32 +18,56 @@ package org.modelix.mps.sync.transformation.modelixToMps.initial
...
+            branch.runRead {
+                val moduleNode =
+                    branch.getRootNode().allChildren.firstOrNull { it.getPropertyValue(BuiltinLanguages.MPSRepositoryConcepts.Module.id) == moduleId }
+                require(moduleNode != null) { "Module node with ID '$moduleId' is not found on the root level." }
+                result.set(moduleNode)
+            }
+            result.get()
+        }.continueWith(linkedSetOf(SyncLock.MODELIX_READ), SyncDirection.NONE) {
+            val entryPoint = it as INode
+            require(entryPoint.isModule()) { "Transformation entry point (Node $entryPoint) must be a Module" }
```
Comment (odzhychko) [inline mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/transformation/modelixToMps/initial/ITreeToSTreeTransformer.kt:58]: nit: This can be checked earlier

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/bindings/ModelBinding.kt [hunk 4]
```diff
@@ -71,7 +72,7 @@ class ModelBinding(val model: SModelBase, branch: IBranch) : IBinding {
 
     override fun deactivate(removeFromServer: Boolean, callback: Runnable?): CompletableFuture<Any?> {
         if (isDisposed) {
-            return CompletableFuture.completedFuture(null)
+            return CompletableFuture<Any?>().completeWithDefault()
```
[the remark this answers is not in the document]
Comment (benedekh) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/bindings/ModelBinding.kt:75]: I think, the implementation of isDone is depends on the JVM we use. That's why I introduced the `completeWithDefault` function`. In GraalVM it's as follows: 
image

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/ModelSyncService.kt [hunk 5]
```diff
@@ -87,27 +90,33 @@ class ModelSyncService : Disposable {
         callback: (() -> Unit),
     ) {
         coroutineScope.launch {
-            logger.info { "disconnecting to server: ${modelClient.baseUrl}" }
-            syncService.disconnectModelServer(modelClient, callback)
-            logger.info { "disconnected server: ${modelClient.baseUrl}" }
+            try {
+                logger.info { "disconnecting to server: ${modelClient.baseUrl}" }
+                syncService.disconnectModelServer(modelClient)
```
Comment (odzhychko) [inline mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/ModelSyncService.kt:95]: thought: `disconnectModelServer` might throw an execption in which case a client in undined state might be still in `activeClients`.

That is why in thet long-term I would like only to have one active client per project.

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/SyncServiceImpl.kt [hunk 6]
```diff
@@ -58,15 +55,10 @@ class SyncServiceImpl : SyncService {
...
-            modelClientV2.init()
-        } catch (e: ConnectException) {
-            logger.warn { "Unable to connect: ${e.message} / ${e.cause}" }
-            throw e
-        }
+        val modelClientV2 = ModelClientV2.builder().url(serverURL.toString()).build()
+        modelClientV2.init()
 
         logger.info { "Connection to $serverURL successful" }
         activeClients.add(modelClientV2)
```
Comment (odzhychko) [inline mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/SyncServiceImpl.kt:64]: Why is not `ModelSyncService.connectModelServer` used to create and keep track of the client?

**Review by odzhychko — CHANGES_REQUESTED** [review]

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/bindings/ModelBinding.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (odzhychko) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/bindings/ModelBinding.kt:75]: But when you finshed a `CompletableFuture` with `null`, then `result` is `NIL`.

```java
    // CompletableFuture
     AltResult NIL = new AltResult((Throwable)null);
     public T get() throws InterruptedException, ExecutionException {
        [...]
        return reportGet(r);
    }
    private static Object reportGet(Object r) throws InterruptedException, ExecutionException {
            [...]
            if ((x = ((AltResult)r).ex) == null) {
                return null;
            }
            [...] 
    }

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt [hunk 7]
```diff
@@ -271,59 +274,78 @@ class ModelSyncGuiFactory : ToolWindowFactory, Disposable {
...
+                moduleModel.addAll(moduleNodes.toList())
+                if (moduleModel.size > 0) {
+                    moduleModel.selectedItem = moduleModel.getElementAt(0)
+                }
+            }
+        }
+
+        @Synchronized
+        private fun callOnlyIfNotFetching(action: suspend () -> Unit) {
+            if (!isFetchingModulesListFromServer.get()) {
```
Comment (odzhychko) [inline mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt:339]: Should the combo boxes be disabled for user input while fetching? And should and indicator be shown in the UI?
Maybe this is out of scope for this PR. If so, could you add a short note in the code about the planned long-term behaviour?

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/bindings/ModelBinding.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (benedekh) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/bindings/ModelBinding.kt:75]: Hmm, it seems to be correct. I'll revert then the default values completion. However, there was a strange corner-case why I had to introduce it.

image

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt (hunk 2 above)
Comment (benedekh) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt:141]: Due to time pressure, I would prefer to skip this aspect and do it when we get there: https://issues.modelix.org/issue/MODELIX-536

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt [hunk 8]
```diff
@@ -136,10 +136,13 @@ class SNodeFactory(
...
+                else -> null
+            }!!
```
[the remark this answers is not in the document]
Comment (benedekh) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt]: fixed in 3ded5600af7eb3a64640d5652eca98fba76484fc

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt (hunk 2 above)
Comment (benedekh) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt:141]: > A comment on why ReferenceLinkFromName can happen.

I'm not sure how that can happen. I just looked for all classes that implement the IReferenceLink interface. However, `GeneratedReferenceLink` is not used here, so I'll add that to make the implementation complete.

(The only missing class is SimpleReferenceLink, but that is only used in our code-based metamodel of the RepositoryConcepts and so.)

Comment (benedekh) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt]: ☝️  is fixed in b79fb24de7633fd5813ba2e45c2b89c706344a88

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/transformation/modelixToMps/initial/ITreeToSTreeTransformer.kt (hunk 3 above)
Comment (benedekh) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/transformation/modelixToMps/initial/ITreeToSTreeTransformer.kt]: fixed in c7fbe666f23b477e94413e7edabeb1f66be3ff28

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/bindings/ModelBinding.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (benedekh) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/bindings/ModelBinding.kt:75]: fixed in a7f49a6770f484ae5ac7312a085345fabcd17dfc

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/SyncServiceImpl.kt [hunk 9]
```diff
@@ -58,15 +55,10 @@ class SyncServiceImpl : SyncService {
...
         // TODO: use JWT here
-        val modelClientV2: ModelClientV2 = ModelClientV2.builder().url(serverURL.toString()).build()
-        try {
-            logger.info { "Connecting to $serverURL" }
-            modelClientV2.init()
-        } catch (e: ConnectException) {
-            logger.warn { "Unable to connect: ${e.message} / ${e.cause}" }
-            throw e
-        }
+        val modelClientV2 = ModelClientV2.builder().url(serverURL.toString()).build()
```
[the remark this answers is not in the document]
Comment (benedekh) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/SyncServiceImpl.kt:52]: fixed in [49ee4cd](https://github.com/modelix/modelix.mps-plugins/pull/73/commits/49ee4cdb9aeeb1cbd46bb798084dbdebe1cf7efa)

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/ModelSyncService.kt (hunk 5 above)
Comment (benedekh) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/ModelSyncService.kt]: fixed in [49ee4cd](https://github.com/modelix/modelix.mps-plugins/pull/73/commits/49ee4cdb9aeeb1cbd46bb798084dbdebe1cf7efa)

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt [hunk 10]
```diff
@@ -271,59 +274,78 @@ class ModelSyncGuiFactory : ToolWindowFactory, Disposable {
...
+                isFetchingModulesListFromServer.set(true)
+
+                coroutineScope.launch {
```
[the remark this answers is not in the document]
Comment (benedekh) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt]: fixed in c13ac8febff9173492f2f6a3a18e137fd4847849

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt [hunk 11]
```diff
@@ -86,6 +87,8 @@ class ModelSyncGuiFactory : ToolWindowFactory, Disposable {
         }
 
         private val logger = KotlinLogging.logger {}
+        private val coroutineScope = CoroutineScope(Dispatchers.Default)
+        private val isFetchingModulesListFromServer = AtomicBoolean()
```
[the remark this answers is not in the document]
Comment (benedekh) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt]: obsolete since c13ac8febff9173492f2f6a3a18e137fd4847849

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt (hunk 7 above)
Comment (benedekh) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt]: Fixed in 71d44f0bf1c3784025660898d612f589af273833.

On the long term we have more UI feedback on the roadmap, see [MODELIX-621](https://issues.modelix.org/issue/MODELIX-621) and [MODELIX-725](https://issues.modelix.org/issue/MODELIX-725).

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt [hunk 12]
```diff
@@ -259,7 +262,7 @@ class ModelSyncGuiFactory : ToolWindowFactory, Disposable {
         private fun triggerRefresh() {
             populateProjectsCB()
             populateConnectionsCB()
-            populateRepoCB()
+            callOnlyIfNotFetching(::populateRepoCB)
```
[the remark this answers is not in the document]
Comment (benedekh) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt:271]: I would like to postpone this question to the future, when we know more specifically how we want to test the frontend/UI of the plugin. At the moment, it's okay that the model-view-controller and closely coupled with each other.

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/ModelSyncService.kt (hunk 5 above)
Comment (odzhychko) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/ModelSyncService.kt]: If `disconnectModelServer`fails, `activeClients` would still contain `modelClient`.

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (odzhychko) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt:271]: Yeah, sure. It was only an `idea`.

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt (hunk 2 above)
Comment (odzhychko) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt]:
```kt
      val reference = when (val referenceLink = it.first) {
                is MPSReferenceLink -> referenceLink.link
                // `referenceLink` is `IReferenceLink`. `IReferenceLink.getSimpleName()` always exists.
                else -> source.concept.referenceLinks.first { refLink -> refLink.name == referenceLink.getSimpleName() }
            }
 ```

You can just use `IReferenceLink.getSimpleName()` directly without type checks.

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (odzhychko) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt]: You can use `kotlinx.coroutines.sync.Mutex` instead of `kotlinx.coroutines.sync.Semaphore`.

> Semaphore with permits = 1 is essentially a [Mutex](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.sync/-mutex/index.html).

See https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.sync/-semaphore/

```kt
        private val mutex = Mutex()

        private fun callOnlyIfNotFetching(action: suspend () -> Unit) {
            coroutineScope.launch {
                if (mutex.tryLock()) {
                    try {
                        setUiControlsEnabled(false)
                        action()
                    } catch (ex: Exception) {
                        logger.error(ex) { "Failed to fetch data from server." }
                    } finally {
                        setUiControlsEnabled(true)
                        mutex.unlock()
                    }
                }
            }
        }
 ```

File: mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt (hunk 2 above)
Comment (benedekh) [reply mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/mps/factories/SNodeFactory.kt]: fixed in 1fdcf251089c424d0847244bb84abde07d71c7a4

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/ModelSyncService.kt (hunk 5 above)
Comment (benedekh) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/ModelSyncService.kt]: Yes, that's true. But the hope would be that we could disconnect it the second time we try. However, it might be a hopeless try.

What would be your expectation in this case? Shall we remove the client from the `connectionsModel`?

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (benedekh) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt]: This gave me exceptions that when we ewere calling mutex.unlock() then the thread was not in the monitor of the mutex. I.e. we called unlock from a different thread, than where we called tryLock(). However, I'll give it a try again and will let you know in a couple of minutes.

[the remark this answers is not in the document]
Comment (benedekh) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt]: Never mind, that was a ReentrantLock instead of a Mutex. Now, it works fine with the Mutex: fefec270781be7272aaad6325b6cad6d15874bce

File: mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/ModelSyncService.kt (hunk 5 above)
Comment (odzhychko) [reply mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/ModelSyncService.kt]: `activeClients` does not exist anymore. I must have looked at some old code.

>  But the hope would be that we could disconnect it the second time we try.
Interesting solution.

**Review by odzhychko — APPROVED** [review]
I am still upset that nothing is automatically tested. This plugin will be difficult to maintain.

But I guess that is nothing that can be added now.
